docs/ds/cartesian-tree.md

```diff
@@ -1,71 +1,71 @@
+It is obvious that each number enters and exits the right chain at most once (or each node exists in the right chain for a continuous period of time). During this process, we can use stack to maintain the nodes on the right chain of the current Cartesian tree. If a node is not on the right chain, we could just pop it off. In this way, each node can enter and exit at most once, and the time complexity is $O(n)$ . The pseudo code is shown as follows:
 
 ```text
-新建一个大小为 n 的空栈。用 top 来标操作前的栈顶，k 来标记当前栈顶。
+Create an empty stack of size n. Use top to denote the top of the stack before the operation, and k to denote the top of the current stack.
 For i := 1 to n
     k := top
-    While 栈非空 且 栈顶元素 > 当前元素 
+    While (stack is not empty) and (top stack element > current element) 
         k--
-    if 栈非空
-        栈顶元素.右儿子 := 当前元素
+    if stack is not empty
+        top stack element.right child := current element
     if k < top
-        当前元素.左儿子 := 栈顶元素
-    当前元素入栈
+        current element.left child := top stack element
+    push current element to stack
     top := k
 ```
 
 ```cpp
 for (int i = 1; i <= n; i++) {
   int k = top;
   while (k > 0 && h[stk[k]] > h[i]) k--;
-  if (k) rs[stk[k]] = i;  // rs代表笛卡尔树每个节点的右儿子
-  if (k < top) ls[i] = stk[k + 1];  // ls代表笛卡尔树每个节点的左儿子
+  if (k) rs[stk[k]] = i;  // rs represents the right child of each node in the Cartesian tree
+  if (k < top) ls[i] = stk[k + 1];  // ls represents the left child of each node in the Cartesian tree
   stk[++k] = i;
   top = k;
 }
 ```
 
-## 笛卡尔树与 Treap
+## Cartesian tree and Treap
 
-谈到笛卡尔树，很容易让人想到一种家喻户晓的结构——Treap。没错，Treap 是笛卡尔树的一种，只不过 w 的值完全随机。Treap 也有线性的构建算法，如果提前将元素排好序，显然可以使用上述单调栈算法完成构建过程，只不过很少会这么用。
+When it comes to Cartesian trees, it is easy to think of a well-known structure - [Treap](https://en.wikipedia.org/wiki/Treap). Yes, Treap is a kind of Cartesian tree, but the value of w is completely random. Treap also has a linear construction algorithm. If you sort the elements in advance, obviously you can use the above monotonic stack algorithm to complete the construction process, but it is rarely used.
 
-## 例题
+## Sample problem
 
-HDU 1506 最大子矩形
+[HDU 1506 Largest Rectangle in a Histogram](http://acm.hdu.edu.cn/showproblem.php?pid=1506)
 
-> 题目大意：n 个位置，每个位置上的高度是 $h_i$ ，求最大子矩阵。举一个例子，如下图：
+> The general idea of the problem: there are n positions, the height of each which is $h_i$ , find the largest sub-matrix. Example is shown below:
 >
 > ![eg](./images/cartesian-tree3.jpeg)
 >
-> 阴影部分就是图中的最大子矩阵。
+> The shaded part is the largest rectangle in a histogram in the figure.
 
-这道题你可 DP，可单调栈，但你万万没想到的是它也可以笛卡尔树！具体地，我们把下标作为键值 $k$ ， $h_i$ 作为键值 $w$ 满足小根堆性质，构建一棵 $(i,h_i)$ 的笛卡尔树。
+This problem clearly can be solved using DP or monotonous stack, but actually it can also be  solved with a Cartesian tree. Specifically, we take the index as the key $k$ and $h_i$ as the value $w$ to satisfy property of the min heap, and build a Cartesian tree of $(i,h_i)$ .
```

Comment:
```suggestion
This problem clearly can be solved using DP or monotonous stack, but actually it can also be solved with a Cartesian tree. Specifically, we take the index as the key $k$ and $h_i$ as the value $w$ to satisfy property of the min heap, and build a Cartesian tree of $(i,h_i)$ .
```